File: js/src/services/visualizationService.js
/**
 * Visualization service for creating charts - JavaScript version
 * Uses Chart.js for rendering charts
 */

const { CHART_COLORS } = typeof module !== 'undefined' && module.exports
    ? require('../config/constants')
    : window.VNGConstants;

/**
 * Service for creating visualizations using Chart.js
 */
class VisualizationService {
    /**
     * Create a line chart for metric trends
     * @param {string} metricName - Name of the metric
     * @param {number[]} values - Array of values
     * @param {string[]} fileNames - Array of file names
     * @param {boolean[]} flags - Array of flag indicators
     * @param {boolean} showConfidence - Whether to show confidence intervals
     * @returns {Object} Chart.js configuration object
     */
    static createLineChart(metricName, values, fileNames, flags = [], showConfidence = false) {
        const datasets = [{
            label: metricName,
            data: values,
            borderColor: CHART_COLORS[0],
            backgroundColor: CHART_COLORS[0] + '20',
            borderWidth: 2,
            fill: false,
            pointRadius: flags.map(flag => flag ? 8 : 4),
            pointBackgroundColor: flags.map(flag => flag ? '#EF4444' : CHART_COLORS[0]),
            pointBorderColor: flags.map(flag => flag ? '#EF4444' : CHART_COLORS[0]),
            pointBorderWidth: 2
        }];

        // Add trendline if enough points
        if (values.length >= 3) {
            const trendline = this._calculateTrendline(values);
            if (trendline.length > 0) {
                datasets.push({
                    label: 'Trend',
                    data: trendline,
                    borderColor: '#6B7280',
                    backgroundColor: 'transparent',
                    borderWidth: 1,
                    borderDash: [5, 5],
                    fill: false,
                    pointRadius: 0
                });
            }
        }

        return {
            type: 'line',
            data: {
                labels: fileNames,
                datasets: datasets
            },
            options: {
                responsive: true,
                plugins: {
                    title: {
                        display: true,
                        text: `${metricName} Trend`
                    },
                    legend: {
                        display: true
                    },
                    tooltip: {
                        callbacks: {
                            afterLabel: (context) => {
                                const index = context.dataIndex;
                                if (flags[index]) {
                                    return '⚠️ FLAGGED';
                                }
                                return '';
                            }
                        }
                    }
                },
                scales: {
                    y: {
                        beginAtZero: false,
                        title: {
                            display: true,
                            text: 'Value'
                        }
                    },
                    x: {
                        title: {
                            display: true,
                            text: 'File'
                        }
                    }
                }
            }
        };
    }

    /**
     * Create a bar chart for category comparison
     * @param {string} categoryName - Name of the category
     * @param {Object} categoryMetrics - Metrics data for the category
     * @param {string[]} fileNames - Array of file names
     * @returns {Object} Chart.js configuration object
     */
    static createBarChart(categoryName, categoryMetrics, fileNames) {
        const metrics = Object.keys(categoryMetrics);
        const datasets = [];

        fileNames.forEach((fileName, fileIndex) => {
            const data = metrics.map(metric => {
                const metricData = categoryMetrics[metric];
                return metricData.values[fileIndex] || 0;
            });

            datasets.push({
                label: fileName,
                data: data,
                backgroundColor: CHART_COLORS[fileIndex % CHART_COLORS.length],
                borderColor: CHART_COLORS[fileIndex % CHART_COLORS.length],
                borderWidth: 1
            });
        });

        return {
            type: 'bar',
            data: {
                labels: metrics,
                datasets: datasets
            },
            options: {
                responsive: true,
                plugins: {
                    title: {
                        display: true,
                        text: `${categoryName} Comparison`
                    },
                    legend: {
                        display: true
                    }
                },
                scales: {
                    y: {
                        beginAtZero: false,
                        title: {
                            display: true,
                            text: 'Value'
                        }
                    },
                    x: {
                        title: {
                            display: true,
                            text: 'Metric'
                        }
                    }
                }
            }
        };
    }

    /**
     * Create a heatmap visualization (using a table representation)
     * @param {Object} analysisResults - Analysis results
     * @param {Array} fileDataList - List of file data
     * @returns {Object} Heatmap data structure for rendering
     */
    static createHeatmap(analysisResults, fileDataList) {
        const heatmapData = {
            categories: [],
            metrics: [],
            values: [],
            fileNames: fileDataList.map(f => f.name)
        };

        const categories = Object.keys(analysisResults).sort();

        categories.forEach(category => {
            const categoryMetrics = Object.keys(analysisResults[category]).sort();

            categoryMetrics.forEach(metric => {
                heatmapData.categories.push(category);
                heatmapData.metrics.push(`${category}: ${metric}`);
                heatmapData.values.push(analysisResults[category][metric].values);
            });
        });

        return heatmapData;
    }

    /**
     * Calculate simple trendline
     * @private
     * @param {number[]} values - Array of values
     * @returns {number[]} Trendline values
     */
    static _calculateTrendline(values) {
        const n = values.length;
        if (n < 2) return [];

        const xValues = Array.from({length: n}, (_, i) => i);
        const validPoints = [];

        for (let i = 0; i < n; i++) {
            if (values[i] !== null && values[i] !== undefined && !isNaN(values[i])) {
                validPoints.push([xValues[i], values[i]]);
            }
        }

        if (validPoints.length < 2) return [];

        const sumX = validPoints.reduce((sum, [x]) => sum + x, 0);
        const sumY = validPoints.reduce((sum, [, y]) => sum + y, 0);
        const sumXY = validPoints.reduce((sum, [x, y]) => sum + x * y, 0);
        const sumXX = validPoints.reduce((sum, [x]) => sum + x * x, 0);

        const vn = validPoints.length;
        const denominator = vn * sumXX - sumX * sumX;

        if (denominator === 0) return [];

        const slope = (vn * sumXY - sumX * sumY) / denominator;
        const intercept = (sumY - slope * sumX) / vn;

        return xValues.map(x => slope * x + intercept);
    }
}

// Export service
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { VisualizationService };
} else if (typeof window !== 'undefined') {
    window.VNGServices = window.VNGServices || {};
    window.VNGServices.VisualizationService = VisualizationService;
}

File: _python_backup/services/visualization_service.py
"""
Visualization service for creating charts
"""

from typing import Dict, List, Any, Tuple
import plotly.graph_objects as go
from modules.visualizer import render_line_chart as _render_line_chart
from modules.visualizer import render_category_chart as _render_category_chart
from domain.models import AnalysisResult, MetricData


class VisualizationService:
    """Service for creating visualizations"""
    
    @staticmethod
    def create_line_chart(
        metric_name: str,
        values: List[float],
        file_names: List[str]
    ) -> go.Figure:
        """
        Create a line chart for a metric
        
        Args:
            metric_name: Name of the metric
            values: List of values across files
            file_names: List of file names
            
        Returns:
            Plotly figure
        """
        return _render_line_chart(metric_name, values, file_names)
    
    @staticmethod
    def create_category_chart(
        category_name: str,
        metrics_map: Dict[str, MetricData],
        file_names: List[str],
        file_data_list: List[Dict[str, Any]]
    ) -> Tuple[go.Figure, bool]:
        """
        Create a category comparison chart
        
        Args:
            category_name: Name of the category
            metrics_map: Dictionary of metric data
            file_names: List of file names
            file_data_list: List of file data (for legacy compatibility)
            
        Returns:
            Tuple of (Plotly figure, show_disclaimer flag)
        """
        # Convert domain models to dict format for legacy function
        metrics_dict = {
            metric: {
                'values': data.values,
                'flags': data.flags,
                'delta': data.delta,
                'percent_change': data.percent_change,
                'std_dev': data.std_dev
            }
            for metric, data in metrics_map.items()
        }
        
        return _render_category_chart(
            category_name,
            metrics_dict,
            file_names,
            file_data_list
        )
    
    @staticmethod
    def create_line_chart_from_analysis(
        analysis_result: AnalysisResult,
        metric_name: str,
        file_names: List[str]
    ) -> go.Figure:
        """
        Create a line chart from analysis result
        
        Args:
            analysis_result: AnalysisResult domain model
            metric_name: Name of the metric to chart
            file_names: List of file names
            
        Returns:
            Plotly figure
        """
        if metric_name not in analysis_result.metrics:
            raise ValueError(f"Metric {metric_name} not found in category {analysis_result.category}")
        
        metric_data = analysis_result.metrics[metric_name]
        return _render_line_chart(metric_name, metric_data.values, file_names)

